File: backend/app/domain/repositories/task_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional, Dict
from backend.app.domain.entities.task import Task


class ITaskRepository(ABC):
    @abstractmethod
    def get_all(self, project_dir: str) -> List[Task]:
        pass

    @abstractmethod
    def get_by_id(self, project_dir: str, task_id: str) -> Optional[Task]:
        pass

    @abstractmethod
    def save(self, project_dir: str, task: Task) -> Task:
        pass

    @abstractmethod
    def update(self, project_dir: str, task: Task) -> Task:
        pass

    @abstractmethod
    def delete(self, project_dir: str, task_id: str) -> bool:
        pass

    @abstractmethod
    def update_orders(self, project_dir: str, task_orders: List[Dict]) -> bool:
        pass

    @abstractmethod
    def initialize_empty_csv(self, project_dir: str) -> None:
        pass

File: frontend/src/domain/repositories/taskRepository.ts
import { Task, TaskCreate, TaskUpdate } from "../entities/task";

export interface ITaskRepository {
  getAll(projectName: string): Promise<Task[]>;
  getById(projectName: string, id: string): Promise<Task | null>;
  create(projectName: string, task: TaskCreate): Promise<Task>;
  update(projectName: string, id: string, task: TaskUpdate): Promise<Task>;
  delete(projectName: string, id: string): Promise<void>;
  updateOrders(
    projectName: string,
    orders: { id: string; display_order: number }[],
  ): Promise<void>;
  bulkUpdate(
    projectName: string,
    updates: { id: string; start_date?: string; due_date?: string }[],
  ): Promise<void>;
}
